src/adaXT/random_forest/RandomForest.py

```diff
@@ -0,0 +1,355 @@
+        '''
+        Releases the allocated memory. Call when finished using the data,
+        or when the data was copied somewhere else.
+        '''
+        self._shared.close()
+        self._shared.unlink()
+
+
+class RandomForest:
+    '''
+    The Random Forrest
+    '''
+
+    def __init__(
+            self,
+            forrest_type: str,
+            n_estimators: int = 100,
+            criterion: Criteria = Squared_error,
+            bootstrap: bool = True,
+            n_jobs: int = 1,
+            max_samples: int = None,
+            max_features: None = None,
+            max_depth: int = sys.maxsize,
+            impurity_tol: float = 0,
+            min_samples_split: int = 1,
+            min_samples_leaf: int = 1,
+            min_improvement: float = 0,
+            splitter: Splitter | None = None):
+        """
+        Parameters
+        ----------
+        forrest_type : str
+            Classification or Regression
+        n_estimators : int, default=100
+            The number of trees in the forest.
+        criterion : Criteria, default=Squared_errror
+            The criteria function used to evaluate a split
+        bootstrap : bool, default=True
+            Whether bootstrap is used when building trees
+        n_jobs : int, default=1
+            The number of processes used to fit, and predict for the forest, -1 uses all available proccesors
+        max_samples : int, default=None
+            The number of samples drawn when doing bootstrap
+        max_features: int, float, {"sqrt", "log2"}, default=None
+            The number of features used when doing feature-sampling
+        max_depth : int
+            maximum depth of the tree, by default maximum system size
+        impurity_tol : float
+            the tolerance of impurity in a leaf node, by default 0
+        min_samples_split : int
+            the minimum amount of samples in a split, by default 1
+        min_samples_leaf : int
+            the minimum amount of samples in a leaf node, by default 1
+        min_improvement: float
+            the minimum improvement gained from performing a split, by default 0
+        splitter : Splitter | None, optional
+            Splitter class if None uses premade Splitter class
+        """
+
+        self.forrest_type = forrest_type
+        self.n_estimators = n_estimators
+        self.criterion = criterion
+        self.bootstrap = bootstrap
+        self.n_jobs = n_jobs if n_jobs != -1 else cpu_count()
+        self.max_samples = max_samples
+        self.max_features = max_features
+        self.max_depth = max_depth
+        self.impurity_tol = impurity_tol
+        self.min_samples_split = min_samples_split
+        self.min_samples_leaf = min_samples_leaf
+        self.min_improvement = min_improvement
+        self.splitter = splitter
+        self.forest_fitted = False
+
+    # Function used to call the fit function of a tree
+    def _build_single_tree(self, tree: DecisionTree):
```

Comment:
```suggestion
    def __build_single_tree(self, tree: DecisionTree):
```
Private function with two underscores